product_multi_barcode_supplierinfo/models/supplierinfo.py

```diff
@@ -0,0 +1,64 @@
+# Copyright 2022 ForgeFlow S.L.
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SupplierInfo(models.Model):
+    _inherit = "product.supplierinfo"
+
+    barcode_id = fields.Many2one(comodel_name="product.barcode")
+
+    barcode = fields.Char(
+        compute="_compute_name",
+        inverse="_inverse_barcode_field",
+        readonly=False,
+        store=True,
+    )
+
+    def _compute_name(self):
+        for rec in self:
+            if rec.barcode_id:
+                rec.barcode = rec.barcode_id.name
+
+    def _inverse_barcode_field(self):
+        if not self.barcode_id:
+            self.barcode_id = self.barcode_id.create({"name": self.barcode})
+        else:
+            self.barcode_id.name = self.barcode
+        self.barcode_id.supplier_id = self.name
+        self.barcode_id.product_tmpl_id = self.product_tmpl_id
+        if self.product_tmpl_id.product_variant_count > 1:
+            self.barcode_id.product_id = self.product_id
+
+    @api.constrains("barcode")
+    def check_barcode(self):
+        for rec in self:
+            if (
+                rec.product_tmpl_id.product_variant_count > 1
+                and rec.barcode
+                and (
+                    not rec.product_id
+                    or rec.product_id == rec.product_tmpl_id.product_variant_id
```

Comment:
why this last condition? `rec.product_id == rec.product_tmpl_id.product_variant_id`

Reply:
to avoid assigning the barcode to the template variant, since it is not an actual variant  with assigned attributes